### PR TITLE
Bugfix: removeFileProtocol adds `/` at the end

### PR DESCRIPTION
**Pull request by gabyx**
`removeFileProtocol` adds a `/` at the end (why??)
resulting in the enhance function in src/render-enhancers/embedded-local-images.ts
not working properly -> file cannot be found.

**What is the sideeffect of removing this magically strange `/`** at the end which should solve this problem??

### CODE REVIEW

**Comment by gabyx** [comment]
Ahh the slash should be at the end....

**Comment by gabyx** [comment]
Ok I think It should behave as the following:

```typescript
export function removeFileProtocol(filePath: string): string {
  return filePath
    .replace(/^(file|vscode-resource|vscode-webview-resource)\:\/\//, "")
}
```
we simply remove the prefix, not more slashes, this works on linux too and on windows its the same 
as bevore.

**Comment by shd101wyy** [comment]
The vscode will use `vscode-resource://file///a/b/c.txt` like file path. There is an extra `file///` which should be replaced with `/`.

**Comment by gabyx** [comment]
Ahh, ok my fix is not yet good:
Win:
- `"file:///F:\Repositories\TechnicalMarkdown\sessions\files\AffineKoordinatentransformation.svg"
Unix:
- `"file:///TechnicalMarkdown\sessions\files\AffineKoordinatentransformation.svg"

and the cases with the `vscode-resource://file///`

Thanks for hint. I try to imporve this behavior

**Comment by gabyx** [comment]
Why you want to transform the path 
`vscode-resource://file///a/b/c.txt`
into 
`/a/b/c.txt`
will that be readable after in e.g. `readFile` (what does happen on Windows?)  which is the **primary** usecase where this `resolveFilePath`  happens.

When do we have `vscode-resource://file///...` links ? in image urls?

**Comment by gabyx** [comment]
Please see the changes. Thats the fix you are proposing.

**Comment by shd101wyy** [comment]
`vscode-resource://file///...` happens in the vscode-markdown-preview-enhanced extension.  
Please check this issue:  https://github.com/shd101wyy/vscode-markdown-preview-enhanced/issues/348

**Comment by shd101wyy** [comment]
I just tried your code and it seems to be wrong in some cases:

For example

```javascript
removeFileProtocol("file://a/b/c.txt")
// returns 'a/b/c.txt' 
// but should be '/a/b/c.txt'
```

Thanks

**Comment by gabyx** [comment]
I dont think its wrong for file://

Thats a good question why should it be an absolute path. in your ex: exactly this path goes to readFile which will fail because /a/b/c does not exist. the correct path for your example is a/b/c and dep. on the current path (think its the workspace folder) of the markdown it should exist, maybe I am wrong and the web industry has here some other notion??


if its an absolute path I (as I debugged and is what I ve seen ob linux, win has a drive letter on front) I will receive file:///a/b/c or file:///C:/a/b/c in the enhance function
which results in /a/b/c or C:/a/b/c which is correct. IMO

Could you pleae give me some image url markdown syntax and the url what you would expect? Thx I will lool at the bug later! Thanks

Von meinem iPhone gesendet

**Comment by gabyx** [comment]
Also here 

https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L452
https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L450

this seems crude why you add an additional / ??

and then all maybe additional slashes (unix) are  trimmed off right after here

https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L452

This seems strange and doesnt look right? or does it?


https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L452
https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L452https://github.com/shd101wyy/mume/blob/026cdd4bd601e75793411b07ff9c7fbd616c69d3/src/utility.ts#L452
Von meinem iPhone gesendet

**Comment by gabyx** [comment]
Ok that with file:/// is correct as I read on wiki.

**Comment by gabyx** [comment]
Ok it seems that stripping of all slashes is correct! Sorry about my unqualified post above. I will fix the bugfix because the problem I think was only the slash witch was added at the end! Thx for the hints

**Comment by gabyx** [comment]
The fix will then be much easier and I can do this on Thuesday next week

File: src/utility.ts [hunk 1]
```diff
@@ -459,17 +459,24 @@ export function addFileProtocol(
...
+  return filePath.replace(regex, (m, isVSCode, rest) => {
+    if (isVSCode) {
+      return "/" + rest;
+    } else {
+      if (process.platform === "win32") {
+        // trim off all excessive '/'
+        return rest.replace(/^\/*/, "");
+      } else {
+        // trim off all excessive '/' except one
+        return rest.replace(/^\/\/*/, "");
```
[the remark this answers is not in the document]
Comment (gabyx) [reply src/utility.ts:476]: Jeah, but wait I push a new fix, will be easier and more  correct than the one now!

**Comment by gabyx** [comment]
Please see the following commit.

File: src/utility.ts [hunk 2]
```diff
@@ -459,17 +459,24 @@ export function addFileProtocol(
...
-  }
+  // See https://regex101.com/r/YlpEur/8/
+  // - "file://///a///b//c/d"                   ---> "a///b//c/d"
+  // - "vscode-resource://///file///a///b//c/d" ---> "file///a///b//c/d"
+  const regex = /^(?:(?:file|(vscode)-(?:webview-)?resource|vscode--resource):\/+)(.*)/m;
+
+  return filePath.replace(regex, (m, isVSCode, rest) => {
+    if (isVSCode) {
+      // For vscode urls -> Remove `://file/...` host
+      rest = rest.replace(/^file\/+/, "/");
```
Comment (shd101wyy) [inline src/utility.ts:470]: Hi @gabyx , I feel this line should be `rest = rest.replace(/^file\/+/, "");`?

Comment (shd101wyy) [reply src/utility.ts:470]: For example, on Windows vscode we will get `vscode-resource://file///C:/a/b/c.txt`, and we want to get `C:/a/b/c.txt` in the end but your code now will give us `/C:/a/b/c.txt`. Thank you.

Comment (gabyx) [reply src/utility.ts:470]: Haha correct :-) Still buggy. Wait I test this function again...

**Comment by gabyx** [comment]
@shd101wyy Jeah you can add me as a collaborator, thx.

**Comment by shd101wyy** [comment]
@gabyx invitations sent. Thank you ;)

File: src/utility.ts (hunk 2 above)
Comment (gabyx) [reply src/utility.ts]: OK now IMO its correct.

**Comment by gabyx** [comment]
Now it should be correct

**Pull request merged**